recipes-openxt/openxt/openxt-keymanagement/key-functions

```diff
@@ -0,0 +1,551 @@
+        dialog --ok-label Continue --msgbox \
+            "ERROR: Keystore does not appear to be available!" 0 0 1>&3
+        return 1;
+    fi
+
+    while [ 1 ] ; do
+        local capture="${KEYDIR}/${key}"
+        dialog --cancel-label Previous --insecure --passwordbox \
+            "Please enter a new password for ${request}" 0 50 "" 2>${capture} 1>&3
+
+        local ret=$?
+        if [ ${ret} != 0 ]; then
+            rm -f ${capture}
+            exec 3>&-
+            return ${ret}
+        fi
+
+        local repeat=$(mktemp -p ${KEYDIR})
+        dialog --cancel-label Previous --insecure --passwordbox \
+            "Please enter the same password to confirm" 0 50 "" 2>${repeat} 1>&3
+
+        local ret=$?
+        if [ ${ret} != 0 ]; then
+            rm -f ${capture}
+            rm -f ${repeat}
+            exec 3>&-
+            return ${ret}
+        fi
+
+        local password=$(cat ${capture})
+
+        if [ -z "${password}" ]; then
+            rm -f ${capture}
+            rm -f ${repeat}
+            dialog --colors --ok-label Continue --msgbox \
+                '  \ZbERROR:\ZB Empty password!\n\n    Please try again.' 7 30 1>&3
+            continue
+        fi
+
+        if diff -q ${capture} ${repeat} 2>/dev/null ; then
+            rm -f ${repeat}
+            break
+        fi
+        rm -f ${repeat}
+
+        dialog --ok-label Continue --msgbox \
+            "ERROR: Passwords did not match. Please try again." 0 0 1>&3
+    done
+
+    exec 3>&-
+    echo ${capture}
+}
+
+recovery_prompt() {
+    local try=""
+
+    # Explicit interact on tty1
+    exec 3>/dev/tty1
+
+    if ! keystore_ready; then
+        dialog --ok-label Continue --msgbox \
+            "ERROR: Keystore does not appear to be available!" 0 0 1>&3
+        return 2;
+    fi
+
+    case ${1} in
+        1) try="first" ;;
+        2) try="second" ;;
+        3) try="third" ;;
+        *) try="more than you should have" ;;
+    esac
+
+    local capture="${KEYDIR}/recovery"
+    dialog --no-cancel --insecure --colors --passwordbox "
+    This is the ${try} attempt.
+
+    Please enter the recovery password " 0 0 2>${capture} 1>&3
+
+    local ret=$?
+    if [ ${ret} != 0 ]; then
+        rm -f ${capture}
+        exec 3>&-
+        return ${ret}
+    fi
+
+    exec 3>&-
+    echo ${capture}
+}
+
+# Seed kernel's CPRNG with some entropy
+# This is a cheap attempt to get the entropy pool to an acceptable
+#   level. If there is a concern then override this function to call
+#   out to an external entropy source.
+seed_entropy() {
+    local seeks="5 72 96 128 51 83 112 12 31 205"
+
+    if [[ ! -e /dev/sda ]]; then
+        return 1
+    fi
+
+    for seek in $seeks; do
+        local entropy=$(cat /proc/sys/kernel/random/entropy_avail)
+        if [[ $entropy -gt 128 ]]; then
+            return 0;
+        fi
+
+        dd if=/dev/sda of=/dev/null bs=1M count=1 seek=${seek} 2>/dev/null
+        sync
+        sleep 2
```

Comment:
AFAIK, by itself, sleeping doesn't generate entropy. It just waits and hopes interrupts and stuff will happen during the sleep. Since 2 seconds * x can add up to a pretty long time, I'm just suggesting we tell the user to generate interrupts during the sleeps by using the keyboard/mouse.
Ultimately, I agree, we should have a proper mechanism like the one you're suggesting, but that can be a separate PR.